### PR TITLE
Cmd ps mode

### PR DESCRIPTION
**Pull request by ajayparida**
Required changes for setting power save mode(wmm/legacy).

### CODE REVIEW

File: drivers/wifi/nrf700x/zephyr/src/zephyr_ps.c [hunk 1]
```diff
@@ -88,16 +88,16 @@ int wifi_nrf_set_power_save_mode(const struct device *dev,
...
 	}
 
-	if (ps_mode_params->mode) {
+	if (ps_mode_params->mode == WIFI_PS_MODE_WMM) {
 		/*WMM mode */
 		status = wifi_nrf_fmac_set_uapsd_queue(rpu_ctx_zep->rpu_ctx,
 						       vif_ctx_zep->vif_idx,
-						       0xF);
+					       WIFI_PS_MODE_WMM_UAPSD_QUEUES);
 	} else {
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_ps.c:96]: Indentation is off.

File: drivers/wifi/nrf700x/zephyr/src/zephyr_ps.c [hunk 2]
```diff
@@ -88,16 +88,16 @@ int wifi_nrf_set_power_save_mode(const struct device *dev,
...
 		/* Legacy mode*/
 		status = wifi_nrf_fmac_set_uapsd_queue(rpu_ctx_zep->rpu_ctx,
 						       vif_ctx_zep->vif_idx,
-						       0);
+					       WIFI_PS_MODE_LEGACY_UAPSD_QUEUES);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_ps.c:100]: Indentation is off.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 3]
```diff
@@ -2952,6 +2952,82 @@ enum wifi_nrf_status wifi_nrf_fmac_set_power_save(void *dev_ctx,
...
+							 unsigned char if_idx,
+							 struct wifi_ps_config *ps_cfg)
+{
+	enum wifi_nrf_status status = WIFI_NRF_STATUS_FAIL;
+	struct nrf_wifi_umac_cmd_get_power_save_cfg *get_ps_cfg_cmd = NULL;
+	struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx = NULL;
+
+	fmac_dev_ctx = dev_ctx;
+
+	get_ps_cfg = wifi_nrf_osal_mem_zalloc(fmac_dev_ctx->fpriv->opriv,
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c]: Is this supposed to be get_ps_cfg**_cmd** ?  If yes how did this compile?

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 4]
```diff
@@ -2952,6 +2952,82 @@ enum wifi_nrf_status wifi_nrf_fmac_set_power_save(void *dev_ctx,
...
+					      sizeof(*get_ps_cfg));
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c]: Is this supposed to be get_ps_cfg**_cmd** ?

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 5]
```diff
@@ -2952,6 +2952,82 @@ enum wifi_nrf_status wifi_nrf_fmac_set_power_save(void *dev_ctx,
...
+
+	if (!get_ps_cfg) {
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c]: Is this supposed to be get_ps_cfg**_cmd** ?

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 6]
```diff
@@ -2952,6 +2952,82 @@ enum wifi_nrf_status wifi_nrf_fmac_set_power_save(void *dev_ctx,
...
+	struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx = NULL;
+
+	fmac_dev_ctx = dev_ctx;
+
+	set_ps_mode_cmd = wifi_nrf_osal_mem_zalloc(fmac_dev_ctx->fpriv->opriv,
+					      sizeof(*set_ps_cmd));
+
+	if (!set_ps_mode_cmd) {
+		wifi_nrf_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+				      "%s: Unable to allocate memory\n", __func__);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c]: Generic comment: Please put each parameter on a separate line (i.e. in this case __func__ can be moved to the next line).

File: drivers/wifi/nrf700x/zephyr/src/zephyr_ps.c [hunk 7]
```diff
@@ -60,3 +60,88 @@ int wifi_nrf_set_power_save(const struct device *dev,
...
+int wifi_nrf_get_power_save_config(const struct device *dev,
+				   struct wifi_ps_config *ps_config)
+{
+	enum wifi_nrf_status status = WIFI_NRF_STATUS_FAIL;
+	struct wifi_nrf_ctx_zep *rpu_ctx_zep = NULL;
+	struct wifi_nrf_vif_ctx_zep *vif_ctx_zep = NULL;
+	struct wifi_ps_config ps_config_info;
+	int ret = -1;
+
+	if (!dev || !params) {
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_ps.c]: params ??

File: drivers/wifi/nrf700x/zephyr/src/zephyr_ps.c [hunk 8]
```diff
@@ -60,3 +60,88 @@ int wifi_nrf_set_power_save(const struct device *dev,
...
+
+int wifi_nrf_set_power_save_mode(const struct device *dev,
+				 struct wifi_ps_mode_params *ps_mode_params)
+{
+	enum wifi_nrf_status status = WIFI_NRF_STATUS_FAIL;
+	struct wifi_nrf_ctx_zep *rpu_ctx_zep = NULL;
+	struct wifi_nrf_vif_ctx_zep *vif_ctx_zep = NULL;
+	int ret = -1;
+
+	if (!dev || !params) {
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_ps.c]: params??

**Review by sachinthegreen — CHANGES_REQUESTED** [review]
Please add a description for the PR.

File: drivers/wifi/nrf700x/zephyr/inc/zephyr_ps.h [hunk 9]
```diff
@@ -17,6 +17,13 @@
 
 #include "osal_api.h"
 
+enum wifi_ps_mode_uapsd_queues {
+	WIFI_PS_MODE_LEGACY_UAPSD_QUEUES = 0,
+	WIFI_PS_MODE_WMM_UAPSD_QUEUES = 0xF,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/inc/zephyr_ps.h:22]: These should be defined in FMAC header file.

File: drivers/wifi/nrf700x/zephyr/src/zephyr_ps.c [hunk 10]
```diff
@@ -60,3 +60,52 @@ int wifi_nrf_set_power_save(const struct device *dev,
...
+	rpu_ctx_zep = vif_ctx_zep->rpu_ctx_zep;
+
+	if (!rpu_ctx_zep) {
+		LOG_ERR("%s: rpu_ctx_zep is NULL\n", __func__);
+		goto out;
+	}
+
+	if (ps_mode_params->mode == WIFI_PS_MODE_WMM) {
+		/*WMM mode */
+		status = wifi_nrf_fmac_set_uapsd_queue(rpu_ctx_zep->rpu_ctx,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_ps.c:93]: Though not in this commit, API looks tied to Linux style, can we use LEGACY/WMM and internally to UMAC use uapsd_queues? If its too much rework then ignore.

**Review by krish2718 — CHANGES_REQUESTED** [review]

**Comment by krish2718** [comment]
You also need to squash the commits.

**Review by sr1dh48r — APPROVED** [review]

**Review by krish2718 — APPROVED** [review]

**Comment by ajayparida** [comment]
> Please add a description for the PR.

Description aded

**Review by sachinthegreen — APPROVED** [review]

**Pull request merged**